src/java.base/share/classes/sun/security/ssl/SSLEngineImpl.java

```diff
@@ -911,13 +912,7 @@ public SSLSession getSession() {
 
     @Override
     public SSLSession getHandshakeSession() {
-        engineLock.lock();
```

Comment:
I'm not very sure of this update. By removing the enginLock on socket/engine level here,  is it possible there are two threads that one read the handshake session and another on write?

Reply:
`engineLock` was used to here to make sure that `conContext.handshakeContext` is not null before accessing it. The problem is that the `handshakeContext` is modified (set to null) by the `TransportContext` class. So using a lock in SSLEngineImpl or SSLSocketImpl doesn't protect `handshakeContext` from becoming null between the check and use.

The actual SSLSession object was never protected from concurrent modification (unless it has its own lock.)

When reviewing the code to answer this, I noticed that I needed to update `SSLEngineImpl.getHandshakeApplicationProtocol()` so there is a new commit.

Reply:
I had a search of the value assignment of handshakeSession.  Here is one example of the calling stack:
SSLEngineImpl.DelegatedTask.run()->TransportContext.dispatch()->SSLHandshake().consume()->HandshakeConsumer.consume()->SSLExtensions.product()->ServerHello.produce()->shc.handshakeSession = session

The engineLock was placed on SSLEngineImpl.DelegatedTask.run() and released after complete the job, which means the value assignment of handshakeSession is synchronized.

The locks in SSLEngineImpl and SSLSocketImpl are used for operations synchronization so that other classes may not need additional locks in most cases.

Reply:
I see what you're saying with respect to SSLEngineImpl. It looks like all of the public methods are synchronized with `engineLock` so I shouldn't have made any changes there. I will revert them.

Are you ok with the changes in SSLSocketImpl?

Reply:
I may update both SSLSocketImpl and SSLEngineImpl, as SSLSocketImpl uses the locks similar to SSLEngineImpl.

Reply:
Yes, socket close is a headache problem for me.

Reply:
There's a bit of a history with SSLSocket closures since the new handshaker was brought into JDK11.  Some of it dealt with synchronization, others with properly handling full vs. half-duplex closes.  You may want to look up some of the bug history on changes surrounding the SSLSocket closures just to make sure we don't undo any fixes that were previously integrated.  As Xuelei's comment indicates, this is a delicate part of the handshaker.

Reply:
I reverted the changes to SSLEngineImpl and looked at the history of SSLSocketImpl and TransportContext but didn't see anything that I would be undoing with this change.

Reply:
Sorry for the delay on any updates here. 

I updated this branch and verified the tests still pass. I ran jdk_security3 tests from test/jdk/TEST.groups. Is there anything else I should do to test this change?

Reply:
The only lock I added was in `TransportContext` to synchronize access to the `handshakeContext` field, but I understand your reluctance to make any changes with locks. The problem is that SSLSocketImpl tries to access `conContext.handshakeContext.negotiatedProtocol` after TransportContext has set handshakeContext to null.

TransportContext also has a `protocolVersion` field. Is it possible to just use that instead?

Reply:
> TransportContext also has a `protocolVersion` field. Is it possible to just use that instead?

Did you mean a change in duplexCloseOutput() like the following?

```
-           // The protocol version may have been negotiated.
-           ProtocolVersion pv = conContext.handshakeContext.negotiatedProtocol;
+           // The protocol version may have been negotiated.  The 
+           // conContext.handshakeContext.negotiatedProtocol is not used as there
+           // may be a race to set it to null.
+          ProtocolVersion pv = conContext.protocolVersion;
            if (pv == null || (!pv.useTLS13PlusSpec())) {
                hasCloseReceipt = true;
            }       
```

I like the idea as it looks like a safe update in the current implementation.

Reply:
Yes, that is what I was thinking.I will make the change and test it out. Thanks.